### ISSUE TITLE
Installation / Import errors with the package

### ISSUE THREAD

**Issue by rottenstea** [maljovec/pyerg#27]
Hi, I am currently trying to install the package, however I can't seem to find a way around some error messages. I tried two different approaches and got two different errors from them:

**Approach 1: pip install nglpy**

When trying to install the package using this command, I get a quite long error message that the setup-wheel cannot be built and that there seems to be a problem with the graph structure:
```
creating build/temp.linux-x86_64-cpython-39
      gcc -pthread -B /home/alena/anaconda3/compiler_compat -Wno-unused-result -Wsign-compare -DNDEBUG -O2 -Wall -fPIC -O2 -isystem /home/alena/anaconda3/include -I/home/alena/anaconda3/include -fPIC -O2 -isystem /home/alena/anaconda3/include -fPIC -I/home/alena/anaconda3/include/python3.9 -c GraphStructure.cpp -o build/temp.linux-x86_64-cpython-39/GraphStructure.o -std=c++11 -O3 -march=native
      In file included from GraphStructure.cpp:36:
      GraphStructure.h:77:64: error: ‘std::string’ has not been declared
         77 |   GraphStructure(std::vector<T> &Xin, int rows, int cols, std::string graph,
            |                                                                ^~~~~~
      GraphStructure.h:173:34: error: ‘std::string’ has not been declared
        173 |                             std::string type, T beta, int &kmax,
            |                                  ^~~~~~
      GraphStructure.cpp:46:6: error: no declaration matches ‘void GraphStructure<T>::compute_neighborhood(std::vector<int>&, std::string, T, int&, bool)’
         46 | void GraphStructure<T>::compute_neighborhood(std::vector<int> &edgeIndices,
            |      ^~~~~~~~~~~~~~~~~
      In file included from GraphStructure.cpp:36:
      GraphStructure.h:172:8: note: candidate is: ‘void GraphStructure<T>::compute_neighborhood(std::vector<int>&, int, T, int&, bool)’
        172 |   void compute_neighborhood(std::vector<int> &edgeIndices,
            |        ^~~~~~~~~~~~~~~~~~~~
      GraphStructure.h:51:7: note: ‘class GraphStructure<T>’ defined here
         51 | class GraphStructure
            |       ^~~~~~~~~~~~~~
      GraphStructure.cpp:157:1: error: no declaration matches ‘GraphStructure<T>::GraphStructure(std::vector<T>&, int, int, std::string, int, T, std::vector<int>&, bool)’
        157 | GraphStructure<T>::GraphStructure(std::vector<T> &Xin, int rows, int cols,
            | ^~~~~~~~~~~~~~~~~
      In file included from GraphStructure.cpp:36:
      GraphStructure.h:77:3: note: candidate is: ‘GraphStructure<T>::GraphStructure(std::vector<T>&, int, int, int, int, T, std::vector<int>&, bool)’
         77 |   GraphStructure(std::vector<T> &Xin, int rows, int cols, std::string graph,
            |   ^~~~~~~~~~~~~~
      GraphStructure.h:51:7: note: ‘class GraphStructure<T>’ defined here
         51 | class GraphStructure
            |       ^~~~~~~~~~~~~~
      error: command '/usr/bin/gcc' failed with exit code 1
```
I can post the entire error message, if required.

**Approach 2: cloning the git-repository and using the commands "make" and "pip install ."**

With this approach I can actually install the package:
```
(base) alena@alena-ThinkPad-T480s:~/Downloads/nglpy-master$ make
swig -v -python -c++ -o ngl_wrap.cpp ngl.i
Language subdirectory: python
Search paths:
   ./
   ./swig_lib/python/
   /usr/share/swig4.0/python/
   ./swig_lib/
   /usr/share/swig4.0/
Preprocessing...
Starting language-specific parse...
Processing types...
C++ analysis...
Processing nested classes...
Generating wrappers...
mv ngl.py nglpy/ngl.py
(base) alena@alena-ThinkPad-T480s:~/Downloads/nglpy-master$ pip install .
Processing /home/alena/Downloads/nglpy-master
  Preparing metadata (setup.py) ... done
Requirement already satisfied: scipy<2.0,>=1.3.2 in /home/alena/anaconda3/lib/python3.9/site-packages (from nglpy==1.1.2) (1.7.3)
Requirement already satisfied: numpy<2.0,>=1.17.4 in /home/alena/anaconda3/lib/python3.9/site-packages (from nglpy==1.1.2) (1.21.5)
Requirement already satisfied: scikit-learn<1.0,>=0.21.2 in /home/alena/anaconda3/lib/python3.9/site-packages (from nglpy==1.1.2) (0.24.2)
Requirement already satisfied: threadpoolctl>=2.0.0 in /home/alena/anaconda3/lib/python3.9/site-packages (from scikit-learn<1.0,>=0.21.2->nglpy==1.1.2) (2.2.0)
Requirement already satisfied: joblib>=0.11 in /home/alena/anaconda3/lib/python3.9/site-packages (from scikit-learn<1.0,>=0.21.2->nglpy==1.1.2) (1.1.0)
Building wheels for collected packages: nglpy
  Building wheel for nglpy (setup.py) ... done
  Created wheel for nglpy: filename=nglpy-1.1.2-cp39-cp39-linux_x86_64.whl size=169093 sha256=b3e9a2d23e2c8966668d744e0f620ea6b46310b61a3047a7fe28eb7c1e25f418
  Stored in directory: /home/alena/.cache/pip/wheels/14/91/27/98d591538d4cfd61565842c04a1e4e5da9bad593f9ae9997c8
Successfully built nglpy
Installing collected packages: nglpy
Successfully installed nglpy-1.1.2
```
However, when I try testing it in a python console, I get the following error message:
`ImportError: cannot import name '_ngl' from partially initialized module 'nglpy' (most likely due to a circular import) (/home/alena/Downloads/nglpy-master/nglpy/__init__.py)`

I am using Ubuntu 22.04 LTS and python 3.9.

**Comment by dmaljovec** [comment]
That’s not a great experience. Let me try to take a look this weekend. I am not sure the state of it, but I did start a branch to excise the C code out of this project: https://github.com/maljovec/nglpy/tree/pure_python

It might not be as performant, but it might fit your needs in the short term while I figure out what is going wrong here.

**Comment by dmaljovec** [comment]
I just published some wheel packages for both MacOS and manylinux. I am testing the manylinux one now...

**Comment by dmaljovec** [comment]
I tested this on a generic `ubuntu:22.04` docker image where the default version for python 3 is 3.10. It works now whereas before publishing the wheels, I got a similar error to yours.

I am also running a test against the Debian-based `python:3.9` image, and it seems to be installing the wheel there as well no problem.

For testing, I am just doing `pip install nglpy` and then running the test script in the `README`:

```python
import nglpy;
import numpy as np;

aGraph = nglpy.EmptyRegionGraph(max_neighbors=9, relaxed=False, beta=1)
aGraph.build(np.random.rand(100,2))

aGraph.neighbors()
```
